nncf/binarization/cuda/functions_cuda_kernel.cu

```diff
@@ -29,7 +29,7 @@ __device__ bool last_block(int* counter) {
 
 // support only warp size = 32
 template <typename scalar_t>
-__device__ void sum_warp(volatile scalar_t* sharr) {
```

Comment:
I was against FP16 kernels only because of this change. Are you sure that CUDA wouldn't mess up sum due to optimization? Probably this is what leads to excessive relative error in tests.

Reply:
The `sum_warp` function is only used in backward's gradient calculation, but the errors occur even in the forward test cases. Anyway, I made a specialization of the function so that FP32 still uses the volatile-warp-synchronous programming, and FP16 uses explicit warp synchronization (no way around doing this if we want to be sure about sums being correct for FP16).